File: cane-cpp/cane-core/src/runtime_data.cpp
#include "runtime_data_internal.hpp"
#include "format/compile_animation.hpp"
#include "format/compile_model.hpp"
#include "cane/error.hpp"
#include <algorithm>
#include <set>
#include <tuple>

namespace cane {
namespace {
using format::Json;
using format::ModelValidation;
using FactKey = std::tuple<TextureResourceKind, std::string, std::string>;
FactKey key(const DecodedTextureDimensions& fact) {
    return fact.kind == TextureResourceKind::direct ? FactKey(fact.kind, fact.image_id, "") : FactKey(fact.kind, fact.atlas_id, fact.page_id);
}
std::map<FactKey, DecodedTextureDimensions> texture_facts(const std::vector<DecodedTextureDimensions>& facts) {
    std::map<FactKey, DecodedTextureDimensions> result;
    for (const auto& fact : facts) {
        if (fact.kind != TextureResourceKind::direct && fact.kind != TextureResourceKind::atlas_page) throw Error(ErrorCode::invalid_argument, "validateDecodedTextureSizes", "Unknown texture resource kind.", "kind");
        const auto valid_id = [](const std::string& id) { return !id.empty() && format::valid_utf8(id, true); };
        if ((fact.kind == TextureResourceKind::direct && (!valid_id(fact.image_id) || !fact.atlas_id.empty() || !fact.page_id.empty()))
            || (fact.kind == TextureResourceKind::atlas_page && (!valid_id(fact.atlas_id) || !valid_id(fact.page_id) || !fact.image_id.empty()))) throw Error(ErrorCode::invalid_argument, "validateDecodedTextureSizes", "Decoded texture identity does not match its resource kind.", "identity");
        if (fact.width == 0 || fact.height == 0) throw Error(ErrorCode::invalid_argument, "validateDecodedTextureSizes", "Decoded texture dimensions must be positive.", "dimensions");
        if (!result.emplace(key(fact), fact).second) throw Error(ErrorCode::validation_failed, "validateDecodedTextureSizes", "Duplicate decoded texture identity.", "identity");
    }
    return result;
}
}
RuntimeData detail::RuntimeDataAccess::load(Json document, const RuntimeLoadOptions& options, std::vector<RuntimeLoadWarning> warnings, const char* operation, const DirectSizes& inherited_sizes) {
    try {
        auto state = std::make_shared<RuntimeDataState>(); state->index = ModelValidation(document, operation).validate();
        state->document = std::move(document); const auto& source = state->document;
        for (const auto& feature : source["requiredFeatures"]) state->required_features.push_back(feature.get<std::string>());
        state->skeleton = {source["skeleton"]["skeletonId"].get<std::string>(), source["skeleton"]["name"].get<std::string>(), source["skeleton"]["referenceScale"].get<float>()};
        state->warnings = std::move(warnings);
        state->atlas_sources = options.atlas_json; state->allow_unverified_features = options.allow_unverified_features;
        constexpr std::array<const char*, 11> catalogs{"atlases", "images", "audios", "fonts", "bones", "slots", "attachments", "constraints", "skins", "events", "animations"};
        for (std::size_t kind = 0; kind < catalogs.size(); ++kind) {
            const auto* id = kind == 0 ? "atlasId" : kind == 1 ? "imageId" : kind == 2 ? "audioId" : kind == 3 ? "fontId" : "id";
            for (const auto& row : source[catalogs[kind]]) state->catalogs[kind].push_back({row[id].get<std::string>(), row.value("name", row[id].get<std::string>()), row.value("type", "")});
        }
        for (const auto& row : source["bones"]) {
            RuntimeBoneDefinition bone; bone.id = row["id"].get<std::string>(); bone.name = row["name"].get<std::string>();
            if (!row["parentId"].is_null()) { bone.parent_id = row["parentId"].get<std::string>(); bone.parent_index = state->index.catalogs[4].at(*bone.parent_id); }
            bone.setup = {row["x"].get<float>(), row["y"].get<float>(), row["rotation"].get<float>(), row["scaleX"].get<float>(), row["scaleY"].get<float>(), row["shearX"].get<float>(), row["shearY"].get<float>()};
            bone.length = row["length"].get<float>(); bone.transform_mode = format::decode_transform_mode(row["transformMode"].get<std::string>()); state->bones.push_back(std::move(bone));
        }
        for (const auto& supplied : options.atlas_json) if (state->index.catalogs[0].find(supplied.first) == state->index.catalogs[0].end()) throw Error(ErrorCode::invalid_argument, "attachAtlas", "Supplied Atlas ID is not declared by this Runtime document.", "atlasId", supplied.first);
        std::vector<std::string> image_ids; for (const auto& row : source["images"]) image_ids.push_back(row["imageId"].get<std::string>());
        std::unordered_set<std::string> paths;
        ModelValidation validation(source, operation);
        for (const auto* collection : {"atlases", "images", "audios", "fonts"}) for (const auto& row : source[collection])
            if (!ModelValidation::get(row, "path").is_null()) paths.insert(validation.path(row["path"], std::string(collection) + ".path"));
        for (const auto& row : source["atlases"]) {
            const auto id = row["atlasId"].get<std::string>(); const auto found = options.atlas_json.find(id);
            if (found == options.atlas_json.end()) throw Error(ErrorCode::missing_resource, "attachAtlas", "Declared Atlas JSON is required.", {}, id);
            auto atlas = AtlasData::from_json(found->second); if (atlas.atlas_id() != id) validation.fail("atlases.atlasId", "Atlas document identity differs from its reference.");
            atlas.validate_image_ids(image_ids);
            for (const auto& region : atlas.regions()) {
                const auto image = state->index.catalogs[1].at(region.image_id);
                if (ModelValidation::get(source["images"][image], "atlasId") != id) validation.fail("images.atlasId", "Atlas region image belongs to a different resource source.");
            }
            for (const auto& page : atlas.pages()) if (!paths.insert(validation.path(Json(page.image_path), "atlas.pages.image")).second) validation.fail("atlas.pages.image", "Atlas page path collides with another Runtime resource.");
            state->atlases.push_back(std::move(atlas));
        }
        const auto decoded = options.decoded_textures ? texture_facts(*options.decoded_textures) : std::map<FactKey, DecodedTextureDimensions>();
        for (const auto& row : source["images"]) {
            RuntimeImageDefinition image; image.image_id = row["imageId"].get<std::string>(); image.name = row["name"].get<std::string>(); image.mime_type = row["mimeType"].get<std::string>();
            if (ModelValidation::get(row, "atlasId").is_null()) {
                image.path = row["path"].get<std::string>(); const auto fact = decoded.find({TextureResourceKind::direct, image.image_id, ""});
                const auto inherited = inherited_sizes.find(image.image_id);
                image.width = row.contains("width") ? row["width"].get<std::uint32_t>() : fact != decoded.end() ? fact->second.width : inherited != inherited_sizes.end() ? inherited->second.first : 0;
                image.height = row.contains("height") ? row["height"].get<std::uint32_t>() : fact != decoded.end() ? fact->second.height : inherited != inherited_sizes.end() ? inherited->second.second : 0;
                if (image.width == 0 || image.height == 0) throw Error(ErrorCode::missing_resource, operation, "Direct images need declared or decoded dimensions.", "images", image.image_id);
            } else {
                image.atlas_id = row["atlasId"].get<std::string>(); image.atlas_index = state->index.catalogs[0].at(*image.atlas_id);
                const auto& atlas = state->atlases[*image.atlas_index]; const AtlasRegion* region = nullptr;
                try { region = &atlas.region_for_image(image.image_id); }
                catch (const Error& failure) {
                    if (failure.code != ErrorCode::not_found) throw;
                    throw Error(ErrorCode::missing_resource, "attachAtlas", "Image has no region in its declared Atlas.", "images", image.image_id);
                }
                image.region_index = static_cast<std::size_t>(region - atlas.regions().data());
                image.width = region->source_width; image.height = region->source_height;
                if ((row.contains("width") && row["width"].get<std::uint32_t>() != image.width) || (row.contains("height") && row["height"].get<std::uint32_t>() != image.height))
                    throw Error(ErrorCode::validation_failed, "attachAtlas", "Atlas source extent differs from the Runtime image declaration.", "images." + image.image_id, image.image_id);
            }
            state->images.push_back(std::move(image));
        }
        state->model = format::compile_model(source, state->index);
        state->clips = format::compile_animations(source, state->index);
        RuntimeData data(std::move(state)); if (options.decoded_textures) data.validate_decoded_texture_sizes(*options.decoded_textures); return data;
    } catch (const std::bad_alloc&) { throw Error(ErrorCode::resource_limit, operation, "Unable to allocate Runtime data."); }
    catch (const Error& error) {
        // Whole-model validation has one public document field. Retain precise source
        // locations in its diagnostic text, while resource/Atlas/limit errors keep their
        // own structured domains. Instance resource transactions preserve target fields.
        if ((std::string_view(operation) == "loadJson" || std::string_view(operation) == "loadCaneb")
            && error.operation == operation && error.code == ErrorCode::validation_failed && error.field != "requiredFeatures") {
            const auto message = error.field ? *error.field + ": " + error.what() : std::string(error.what());
            throw Error(error.code, error.operation, message, "$", error.entity_id, error.detail);
        }
        throw;
    }
    catch (const Json::exception& error) { throw Error(ErrorCode::validation_failed, operation, error.what(), "$"); }
}
RuntimeData RuntimeData::from_json(std::string_view utf8, const RuntimeLoadOptions& options) { return detail::RuntimeDataAccess::load(format::parse_json(utf8), options, {}, "loadJson"); }
RuntimeData RuntimeData::from_caneb(const std::vector<std::uint8_t>& bytes, const RuntimeLoadOptions& options) {
    auto decoded = format::decode_caneb(bytes); std::vector<RuntimeLoadWarning> warnings;
    for (const auto& warning : decoded.warnings) warnings.push_back({warning.code, warning.operation, warning.section_tag});
    return detail::RuntimeDataAccess::load(std::move(decoded.document), options, std::move(warnings), "loadCaneb");
}
const RuntimeSkeletonInfo& RuntimeData::skeleton() const noexcept { return state_->skeleton; }
const std::vector<RuntimeCatalogEntry>& RuntimeData::catalog(RuntimeCatalogKind kind) const {
    const auto index = static_cast<std::size_t>(kind); if (index >= state_->catalogs.size()) throw Error(ErrorCode::invalid_argument, "getCatalog", "Unknown Runtime catalog kind.", "kind");
    return state_->catalogs[index];
}
const std::vector<RuntimeBoneDefinition>& RuntimeData::bones() const noexcept { return state_->bones; }
const std::vector<RuntimeImageDefinition>& RuntimeData::images() const noexcept { return state_->images; }
const std::vector<AtlasData>& RuntimeData::atlases() const noexcept { return state_->atlases; }
const std::vector<std::string>& RuntimeData::required_features() const noexcept { return state_->required_features; }
const std::vector<RuntimeLoadWarning>& RuntimeData::warnings() const noexcept { return state_->warnings; }
void RuntimeData::validate_decoded_texture_sizes(const std::vector<DecodedTextureDimensions>& decoded) const {
    auto facts = texture_facts(decoded);
    const auto consume = [&](const FactKey& id, std::uint32_t width, std::uint32_t height) {
        const auto found = facts.find(id); if (found == facts.end()) throw Error(ErrorCode::missing_resource, "validateDecodedTextureSizes", "Complete decoded texture catalog is required.", "textures");
        if (found->second.width != width || found->second.height != height) throw Error(ErrorCode::validation_failed, "validateDecodedTextureSizes", "Decoded texture dimensions disagree with the Runtime resource.", "dimensions");
        facts.erase(found);
    };
    for (const auto& image : state_->images) if (image.path) consume({TextureResourceKind::direct, image.image_id, ""}, image.width, image.height);
    for (const auto& atlas : state_->atlases) for (const auto& page : atlas.pages()) consume({TextureResourceKind::atlas_page, atlas.atlas_id(), page.page_id}, page.width, page.height);
    if (!facts.empty()) throw Error(ErrorCode::validation_failed, "validateDecodedTextureSizes", "Decoded catalog contains an undeclared texture.", "textures");
}
}

File: cane-csharp/cane-core/Runtime/RuntimeData.cs
using System;
using System.Collections.Generic;
using Cane.Format;

namespace Cane
{
    public sealed class RuntimeLoadOptions
    {
        /// <summary>Atlas JSON documents, keyed by their stable atlas ID. Resource I/O belongs to the host.</summary>
        public IReadOnlyDictionary<string, string> AtlasJson { get; set; } = new Dictionary<string, string>();
        /// <summary>Optional complete decoded texture catalog. Resolves omitted direct-image dimensions and verifies every direct image and atlas page.</summary>
        public IReadOnlyList<DecodedTextureDimensions>? DecodedTextures { get; set; }
        internal IReadOnlyDictionary<string, DecodedTextureDimensions>? InheritedDirectSizes;
        /// <summary>Retained for source compatibility. Known features load by default; unknown features always fail.</summary>
        public bool AllowUnverifiedFeatures { get; set; }
    }

    public sealed partial class RuntimeData
    {
        internal readonly Json Document;
        internal readonly Dictionary<string, string> AtlasSources = new Dictionary<string, string>(StringComparer.Ordinal);
        internal readonly Dictionary<string, Json> AtlasDocuments = new Dictionary<string, Json>(StringComparer.Ordinal);
        internal readonly Dictionary<string, DecodedTextureDimensions> DirectImageSizes = new Dictionary<string, DecodedTextureDimensions>(StringComparer.Ordinal);
        internal readonly bool AllowUnverifiedFeatures;
        internal readonly BoneData[] BoneData;
        internal readonly SlotData[] SlotData;
        internal readonly AttachmentData[] AttachmentData;
        internal readonly Dictionary<string, int> BoneIndex = new Dictionary<string, int>(StringComparer.Ordinal);
        internal readonly Dictionary<string, int> SlotIndex = new Dictionary<string, int>(StringComparer.Ordinal);
        internal readonly Dictionary<string, int> AttachmentIndex = new Dictionary<string, int>(StringComparer.Ordinal);
        internal readonly Dictionary<string, Json> AnimationDocuments = new Dictionary<string, Json>(StringComparer.Ordinal);
        internal readonly Dictionary<string, Animation.Clip> Clips = new Dictionary<string, Animation.Clip>(StringComparer.Ordinal);
        internal readonly Dictionary<string, Json> SkinDocuments = new Dictionary<string, Json>(StringComparer.Ordinal);
        internal readonly Dictionary<string, Json> EventDocuments = new Dictionary<string, Json>(StringComparer.Ordinal);
        internal readonly Dictionary<string, ImageData> Images = new Dictionary<string, ImageData>(StringComparer.Ordinal);
        internal readonly Json[] Constraints;
        public string SkeletonId { get; }
        public string Name { get; }
        public float ReferenceScale { get; }
        public IReadOnlyList<AnimationInfo> Animations { get; }
        public IReadOnlyList<string> SkinIds { get; }
        public IReadOnlyList<string> RequiredFeatures { get; }
        public IReadOnlyList<string> IgnoredOptionalCanebSections { get; private set; } = Array.AsReadOnly(Array.Empty<string>());

        public static RuntimeData FromJson(string json, RuntimeLoadOptions? options = null) => Load(Json.Parse(json), options ?? new RuntimeLoadOptions());
        public static RuntimeData FromJson(byte[] utf8, RuntimeLoadOptions? options = null) => Load(Json.Parse(utf8), options ?? new RuntimeLoadOptions());
        public static RuntimeData FromCaneb(byte[] bytes, RuntimeLoadOptions? options = null)
        {
            Json document = Caneb.Decode(bytes, out string[] optionalSections);
            RuntimeData data = Load(document, options ?? new RuntimeLoadOptions());
            data.SetLoadWarnings(optionalSections); return data;
        }
        internal void SetLoadWarnings(string[] optionalSections)
        {
            IgnoredOptionalCanebSections = Array.AsReadOnly((string[])optionalSections.Clone());
            var warnings = new RuntimeWarning[optionalSections.Length];
            for (int i = 0; i < warnings.Length; i++) warnings[i] = new RuntimeWarning(optionalSections[i]);
            Warnings = Array.AsReadOnly(warnings);
        }
        public RuntimePlayer CreatePlayer() => new RuntimePlayer(this);

        internal static RuntimeData Load(Json document, RuntimeLoadOptions options)
        {
            try { return new RuntimeData(document, options); }
            catch (RuntimeException) { throw; }
            catch (Exception e) when (e is OverflowException || e is ArgumentException || e is IndexOutOfRangeException)
            { throw new RuntimeException(RuntimeErrorCode.ValidationFailed, "loadJson", "Invalid runtime document: " + e.Message); }
        }

        internal static RuntimeData LoadOverlay(Json document, RuntimeLoadOptions options, RuntimeData source)
        {
            RuntimeData candidate = Load(document, options);
            // Immutable original load diagnostics remain owned metadata, never project JSON.
            candidate.IgnoredOptionalCanebSections = source.IgnoredOptionalCanebSections;
            candidate.Warnings = source.Warnings;
            return candidate;
        }

        private RuntimeData(Json doc, RuntimeLoadOptions options)
        {
            Document = doc;
            AllowUnverifiedFeatures = options.AllowUnverifiedFeatures;
            Validation.Root(doc);
            SkeletonId = doc["skeleton"].S("skeletonId"); Name = doc["skeleton"].S("name"); ReferenceScale = doc["skeleton"].F("referenceScale");
            RequiredFeatures = Array.AsReadOnly(doc["requiredFeatures"].Strings());
            DecodedTextureDimensions[]? decoded = options.DecodedTextures == null ? null : CopyDecodedTextureFacts(options.DecodedTextures, "loadJson");
            var directSizes = new Dictionary<string, DecodedTextureDimensions>(StringComparer.Ordinal);
            if (options.InheritedDirectSizes != null) foreach (var fact in options.InheritedDirectSizes) directSizes.Add(fact.Key, fact.Value);
            if (decoded != null) foreach (var fact in decoded) if (fact.Kind == TextureResourceKind.Direct) directSizes[fact.ImageId!] = fact;
            var bones = doc["bones"].Items;
            BoneData = new BoneData[bones.Count];
            for (int i = 0; i < bones.Count; ++i)
            {
                Json b = bones[i]; Validation.Bone(b);
                string id = Validation.Id(b, "id"), parentId = b.S("parentId");
                int parent = -1;
                if (parentId.Length != 0 && !BoneIndex.TryGetValue(parentId, out parent)) Validation.Missing("bones.parentId", parentId);
                AddId(BoneIndex, id, i); BoneData[i] = new BoneData(id, parent, b);
            }
            var slots = doc["slots"].Items; SlotData = new SlotData[slots.Count];
            for (int i = 0; i < slots.Count; ++i)
            {
                var s = slots[i]; Validation.Slot(s); string id = Validation.Id(s, "id");
                if (!BoneIndex.TryGetValue(s.S("boneId"), out int bone)) Validation.Missing("slots.boneId", s.S("boneId"));
                AddId(SlotIndex, id, i); SlotData[i] = new SlotData(id, bone, s);
            }
            var atlases = new Dictionary<string, Json>(StringComparer.Ordinal);
            var atlasIds = new HashSet<string>(StringComparer.Ordinal);
            foreach (Json a in doc["atlases"].Items)
            {
                Validation.Closed(a, "atlasId path", "atlasId path"); Validation.Path(a.S("path"));
                var id = Validation.Id(a, "atlasId");
                if (!atlasIds.Add(id)) Validation.Fail("Duplicate atlas ID.");
                if (!options.AtlasJson.TryGetValue(id, out var source)) continue;
                Json atlas = Json.Parse(source); Validation.Atlas(atlas);
                AtlasSources.Add(id, source);
                if (atlas.S("atlasId") != id) Validation.Fail("Atlas ID does not match its reference.");
                if (atlases.ContainsKey(id)) Validation.Fail("Duplicate atlas ID."); atlases.Add(id, atlas); AtlasDocuments.Add(id, atlas);
            }
            var imageRows = new Dictionary<string, Json>(StringComparer.Ordinal);
            foreach (var image in doc["images"].Items)
            {
                Validation.Image(image); string id = Validation.Id(image, "imageId");
                if (!image["atlasId"].IsNull && !atlasIds.Contains(image.S("atlasId"))) Validation.Missing("images.atlasId", image.S("atlasId"));
                if (imageRows.ContainsKey(id)) Validation.Fail("Duplicate image ID.");
                imageRows.Add(id, image);
            }
            foreach (var atlas in atlases)
                foreach (Json region in atlas.Value["regions"].Items)
                {
                    string imageId = region.S("imageId");
                    if (!imageRows.TryGetValue(imageId, out Json? image)) Validation.Missing("atlas.regions.imageId", imageId);
                    if (image!.S("atlasId") != atlas.Key) Validation.Fail("Atlas region image belongs to another resource source.");
                }
            var attachments = doc["attachments"].Items; AttachmentData = new AttachmentData[attachments.Count];
            for (int i = 0; i < attachments.Count; ++i)
            {
                var a = attachments[i]; Validation.Attachment(a, BoneIndex, i);
                var id = Validation.Id(a, "id");
                if (!SlotIndex.TryGetValue(a.S("slotId"), out int slot)) Validation.Missing("attachments.slotId", a.S("slotId"));
                AddId(AttachmentIndex, id, i); AttachmentData[i] = new AttachmentData(id, slot, a, BoneIndex);
                if ((a.S("type") == "region" || a.S("type") == "mesh") && !imageRows.ContainsKey(a.S("imageId"))) Validation.Missing("attachments.imageId", a.S("imageId"));
                foreach (Json imageId in a["sequence"]["imageIds"].ArrayOrEmpty) if (!imageRows.ContainsKey(imageId.String)) Validation.Missing("sequence.imageIds", imageId.String);
            }
            foreach (var a in AttachmentData)
            {
                var seen = new HashSet<string>(StringComparer.Ordinal); var current = a;
                while (!current.Raw["link"].IsNull)
                {
                    if (!seen.Add(current.Id)) Validation.Fail("Linked mesh cycle.");
                    string parentId = current.Raw["link"].S("parentMeshId");
                    if (!AttachmentIndex.TryGetValue(parentId, out int parent)) Validation.Missing("link.parentMeshId", parentId);
                    current = AttachmentData[parent]; if (current.Kind != "mesh") Validation.Fail("Linked parent must be a mesh.");
                }
                a.GeometrySource = current;
                current = a;
                while (!current.Raw["link"].IsNull && current.Raw["link"].B("inheritDeform", true)) current = AttachmentData[AttachmentIndex[current.Raw["link"].S("parentMeshId")]];
                a.DeformSourceId = current.Id;
            }
            var skins = new List<string>();
            foreach (Json skin in doc["skins"].Items)
            {
                Validation.Closed(skin, "id name attachments boneIds constraintIds export", "id name attachments boneIds constraintIds export");
                string id = Validation.Id(skin, "id"); if (SkinDocuments.ContainsKey(id)) Validation.Fail("Duplicate skin ID.");
                foreach (Json a in skin["attachments"].Items)
                {
                    Validation.Closed(a, "slotId attachmentId name", "slotId attachmentId name");
                    if (!SlotIndex.TryGetValue(a.S("slotId"), out int slot)) Validation.Missing("skins.slotId", a.S("slotId"));
                    if (!a["attachmentId"].IsNull && (!AttachmentIndex.TryGetValue(a.S("attachmentId"), out int attachment) || AttachmentData[attachment].Slot != slot)) Validation.Fail("Skin attachment must belong to its slot.");
                }
                foreach (var bone in skin["boneIds"].Strings()) if (!BoneIndex.ContainsKey(bone)) Validation.Missing("skins.boneIds", bone);
                skins.Add(id); SkinDocuments.Add(id, skin); CompileSkin(id, skin);
            }
            SkinIds = skins.AsReadOnly();
            foreach (Json e in doc["events"].Items)
            {
                Validation.Closed(e, "id name integerValue stringValue numberValue audioId volume balance", "id name integerValue stringValue numberValue audioId");
                if (!e["integerValue"].IsNull) Validation.SignedInteger(e["integerValue"], "events.integerValue");
                string id = Validation.Id(e, "id"); if (EventDocuments.ContainsKey(id)) Validation.Fail("Duplicate event ID."); EventDocuments.Add(id, e);
            }
            Constraints = doc["constraints"].Items.ToArray();
            foreach (Json constraint in Constraints) Validation.Constraint(constraint, this);
            var animations = new List<AnimationInfo>();
            foreach (Json a in doc["animations"].Items)
            {
                Validation.Animation(a, this);
                string id = Validation.Id(a, "id"); if (AnimationDocuments.ContainsKey(id)) Validation.Fail("Duplicate animation ID.");
                AnimationDocuments.Add(id, a); animations.Add(new AnimationInfo(id, a.S("name"), a.F("duration"), a.F("fps")));
            }
            Animations = animations.AsReadOnly();
            foreach (var atlasId in atlasIds) if (!atlases.ContainsKey(atlasId)) throw new RuntimeException(RuntimeErrorCode.MissingResource, "attachAtlas", "Atlas JSON is required.", entityId: atlasId);
            foreach (var row in imageRows)
            {
                directSizes.TryGetValue(row.Key, out DecodedTextureDimensions? size);
                bool direct = row.Value["atlasId"].IsNull;
                var image = new ImageData(row.Value, direct ? null : atlases[row.Value.S("atlasId")], direct ? size : null); Images.Add(row.Key, image);
                if (direct) DirectImageSizes.Add(row.Key, DecodedTextureDimensions.Direct(row.Key, image.Width, image.Height));
            }
            if (decoded != null) ValidateDecodedTextureFacts(decoded, "loadJson");
            foreach (var animation in AnimationDocuments) Clips.Add(animation.Key, new Animation.Clip(this, animation.Value));
            FeatureUsage.ValidateDeclared(doc);
        }

        private static void AddId(Dictionary<string, int> index, string id, int i)
        { if (index.ContainsKey(id)) Validation.Fail("Duplicate ID: " + id); index.Add(id, i); }
    }

    internal sealed class BoneData
    {
        internal readonly string Id, TransformMode;
        internal readonly int Parent;
        internal readonly float Length;
        internal readonly BoneLocal Setup;
        internal BoneData(string id, int parent, Json b)
        { Id = id; Parent = parent; Length = b.F("length"); TransformMode = b.S("transformMode"); Setup = Local(b); }
        internal static BoneLocal Local(Json b) => new BoneLocal(b.F("x"), b.F("y"), b.F("rotation"), b.F("scaleX", 1), b.F("scaleY", 1), b.F("shearX"), b.F("shearY"));
    }
    internal sealed class SlotData
    {
        internal readonly string Id, Blend;
        internal readonly string? Attachment;
        internal readonly int Bone;
        internal readonly long ZIndex;
        internal readonly Rgb Color;
        internal readonly Rgb? Dark;
        internal readonly float Alpha;
        internal SlotData(string id, int bone, Json s)
        { Id = id; Bone = bone; ZIndex = s.L("zIndex"); Blend = s.S("blendMode"); Attachment = s["attachmentId"].StringOrNull;
          Color = Rgb.Parse(s.S("color", "#ffffff")); Dark = s["darkColor"].IsNull ? (Rgb?)null : Rgb.Parse(s.S("darkColor")); Alpha = s.F("alpha", 1); }
    }
    internal sealed class AttachmentData
    {
        internal readonly struct Influence
        {
            internal readonly int Bone;
            internal readonly float Weight, X, Y;
            internal Influence(int bone, float weight, Point local) { Bone = bone; Weight = weight; X = local.X; Y = local.Y; }
        }
        internal readonly string Id, Kind;
        internal readonly int Slot;
        internal readonly Json Raw;
        internal readonly Rgb Color;
        internal readonly float Alpha;
        internal readonly float[] Vertices, Uvs;
        internal readonly int[] Indices;
        internal readonly int[] InfluenceStarts;
        internal readonly Influence[] Influences;
        internal readonly BoneLocal Setup;
        internal AttachmentData GeometrySource;
        internal string DeformSourceId;
        internal AttachmentData(string id, int slot, Json a, IReadOnlyDictionary<string, int> bones)
        {
            Id = id; Slot = slot; Kind = a.S("type"); Raw = a; Color = Rgb.Parse(a.S("color", "#ffffff")); Alpha = a.F("alpha", 1); Setup = BoneData.Local(a);
            Vertices = a["vertices"].Floats(); Uvs = a["uvs"].Floats(); var source = a["indices"].ArrayOrEmpty; Indices = new int[source.Count];
            for (int i = 0; i < source.Count; ++i) Indices[i] = source[i].Integer;
            var weights = a["weights"].ArrayOrEmpty;
            if (weights.Count == 0) { InfluenceStarts = Array.Empty<int>(); Influences = Array.Empty<Influence>(); }
            else
            {
                InfluenceStarts = new int[weights.Count + 1]; int count = 0;
                for (int i = 0; i < weights.Count; i++) { InfluenceStarts[i] = count; count += weights[i].Items.Count; }
                InfluenceStarts[weights.Count] = count; Influences = new Influence[count]; int cursor = 0;
                for (int vertex = 0; vertex < weights.Count; vertex++)
                    foreach (Json influence in weights[vertex].Items)
                    {
                        float weight = influence.F("weight");
                        // Zero weights still occupy deform offsets, but do not read a bone or bind position during evaluation.
                        if (weight == 0) { Influences[cursor++] = default; continue; }
                        string bone = influence.S("boneId");
                        Point local = influence["x"].IsNull
                            ? Animation.Sampling.ReadAffine(a["bindInverses"][bone]).Transform(Vertices[vertex * 2], Vertices[vertex * 2 + 1])
                            : new Point(influence.F("x"), influence.F("y"));
                        Influences[cursor++] = new Influence(bones[bone], weight, local);
                    }
            }
            GeometrySource = this; DeformSourceId = Id;
        }
    }
    internal sealed class ImageData
    {
        internal readonly TextureDescriptor Texture;
        internal readonly Json? Region;
        internal readonly Geometry.AtlasProjection? Atlas;
        internal readonly int Width, Height;
        internal ImageData(Json image, Json? atlas, DecodedTextureDimensions? decoded = null)
        {
            string id = image.S("imageId");
            if (atlas == null)
            {
                if (decoded != null && ((!image["width"].IsNull && image.I("width") != decoded.Width) || (!image["height"].IsNull && image.I("height") != decoded.Height)))
                    throw new RuntimeException(RuntimeErrorCode.ValidationFailed, "loadJson", "Decoded direct-image dimensions differ from the resource declaration.", "images", id);
                Width = image["width"].IsNull ? decoded?.Width ?? 0 : image.I("width"); Height = image["height"].IsNull ? decoded?.Height ?? 0 : image.I("height");
                if (Width <= 0 || Height <= 0) throw new RuntimeException(RuntimeErrorCode.MissingResource, "loadJson", "Direct images require decoded width and height.", "images", id);
                Texture = new TextureDescriptor(id, image.S("path"), Width, Height); return;
            }
            foreach (var r in atlas["regions"].Items) if (r.S("imageId") == id) { if (Region != null) Validation.Fail("Duplicate atlas region for image."); Region = r; }
            if (Region == null) throw new RuntimeException(RuntimeErrorCode.MissingResource, "loadJson", "Atlas region is missing.", "images", id);
            Atlas = new Geometry.AtlasProjection(Region);
            Width = Region.I("sourceWidth"); Height = Region.I("sourceHeight");
            if ((!image["width"].IsNull && image.I("width") != Width) || (!image["height"].IsNull && image.I("height") != Height)) throw new RuntimeException(RuntimeErrorCode.ValidationFailed, "attachAtlas", "Atlas source dimensions mismatch.", "images." + id, id);
            Json? page = null;
            foreach (var p in atlas["pages"].Items) if (p.S("pageId") == Region.S("pageId")) page = p;
            if (page == null) throw new RuntimeException(RuntimeErrorCode.MissingReference, "loadJson", "Atlas page is missing.");
            Texture = new TextureDescriptor(id, page.S("image"), page.I("width"), page.I("height"), atlas.S("atlasId"),
                page.S("pageId"), Region.S("regionId"), atlas.S("colorSpace"), atlas.S("alphaMode"));
        }
    }
}
